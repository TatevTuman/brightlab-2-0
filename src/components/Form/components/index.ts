export { default as FormItem } from './FormItem/FormItem'
export { default as FormControl } from './FormControl/FormControl'
export { default as FormSubmit } from './FormSubmit/FormSubmit'
