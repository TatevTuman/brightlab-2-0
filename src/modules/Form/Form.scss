.form {
  margin-bottom: 0;

  &-item {
    display: flex;
    align-items: center;
    margin-bottom: 1rem;

    &:last-child {
      margin-bottom: 0 !important;
    }
  }

  &-control {
    width: 100%;

    &[data-no-width='true'] {
      width: unset;
    }
  }

  &-submit {

  }
}
