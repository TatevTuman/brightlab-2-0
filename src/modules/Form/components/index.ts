export { default as FormItem } from './FormItem/FormItem'
export { default as FormControl } from './FormControl'
export { default as FormSubmit } from './FormSubmit'

export type { FormItemProps } from './FormItem/FormItem'
export type { FormControlProps } from './FormControl'
export type { FormSubmitProps } from './FormSubmit'
