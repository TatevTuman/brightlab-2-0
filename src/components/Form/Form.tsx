import React, { memo, forwardRef, RefObject } from 'react'
import {
  useForm,
  SubmitHandler,
  SubmitErrorHandler,
  UnpackNestedValue,
  DeepPartial,
  UseFormMethods,
  FieldName,
  FormProvider
} from 'react-hook-form'
import { Children } from '@types'
import { FormItem, FormSubmit } from './components'
import styles from './Form.module.scss'

export interface FormProps<F> {
  defaultValues?: UnpackNestedValue<DeepPartial<F>>
  children: (props: UseFormMethods<F>) => Children
  onSubmit: SubmitHandler<F>
  onError: SubmitErrorHandler<F>
  ref?: RefObject<HTMLFormElement>
}

/*
  Form component that uses react-hook-form lib to manage form state without React state.
  Forward ref is used in awkward way because of Typescript genetic argument to pass it in useForm hook.
  TODO Research

  Loading state is implemented via FormSubmit component that controls loading field like input in the form
  You can add here any logic you want to improve UX
*/

const Form = <F,>(props: FormProps<F>) => {
  const ForwardRefWrapper = memo(
    forwardRef<HTMLFormElement, FormProps<F>>((props, ref) => {
      const { defaultValues, onSubmit, onError, children } = props
      const useFormMethods = useForm<F>({ defaultValues })

      /* FormSubmit component controls loading field, so we can manage button state with react-hook-form */
      const setFormLoading = () =>
        setTimeout(
          () =>
            useFormMethods.setValue('loading' as FieldName<F>, false, {
              shouldValidate: false,
              shouldDirty: true
            }),
          500
        )

      /* Custom form submit handler */
      const handleFormSubmit: SubmitHandler<F> = async (form, event) => {
        await onSubmit(form, event)
        /* Set FormSubmit loading to false */
        setFormLoading()
      }

      /* Custom form error handler */
      const handleFormError: SubmitErrorHandler<F> = async (errors, event) => {
        await onError(errors, event)
        /* Set FormSubmit loading to false */
        setFormLoading()
      }

      return (
        <FormProvider {...useFormMethods}>
          <form
            className={styles.form}
            ref={ref}
            onSubmit={useFormMethods.handleSubmit<F>(handleFormSubmit, handleFormError)}
          >
            {children(useFormMethods)}
          </form>
        </FormProvider>
      )
    })
  )

  ForwardRefWrapper.displayName = 'Form'

  return <ForwardRefWrapper {...props} />
}

Form.Item = FormItem
Form.Submit = FormSubmit

export default Form
