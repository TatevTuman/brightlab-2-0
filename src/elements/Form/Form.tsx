import React, { memo, forwardRef, ReactElement } from 'react'
import { useForm } from 'react-hook-form'
import { UnpackNestedValue, FieldErrors } from 'react-hook-form/dist/types/form'
import './Form.scss'

interface FormProps {
  defaultValues?: Record<string, any>
  children: JSX.Element | JSX.Element[]
  onSubmit: (data: UnpackNestedValue<Record<string, any>>) => void
  onError?: (errors: FieldErrors<Record<string, any>>) => void
}

const Form: React.ForwardRefExoticComponent<FormProps> = forwardRef<HTMLFormElement, FormProps>((props, ref) => {
  const { defaultValues, onSubmit, onError, children } = props
  const { register, handleSubmit, errors } = useForm({ defaultValues })

  return (
    <form ref={ref} onSubmit={handleSubmit(onSubmit, onError)}>
      {React.Children.map(children, (child: ReactElement) => {
        return child.props.name
          ? React.createElement(child.type, {
              ...{
                ...child.props,
                register: register(child.props.validation),
                error: errors[child.props.name],
                key: child.props.name
              }
            })
          : child
      })}
    </form>
  )
})

export default memo(Form)
