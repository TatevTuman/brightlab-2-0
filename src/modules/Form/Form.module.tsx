import React from 'react'
import { ControllerRenderProps } from 'react-hook-form'
import { LoadableComponent } from '@loadable/component'
import { withLoadableFallback, WithLoadableFallbackOptions } from '@hocs'
import { Autocomplete, AutocompleteProps, Select, SelectProps } from '@components'
import { Input, InputProps, Checkbox, CheckboxProps } from '@elements'
import type { FormProps, FormControlProps } from '@modules'

import { FormControl, FormItem, FormSubmit } from './components'

/* We don't need to pass these props to form component anymore because form controls them */
export type FormModuleComponentControlledFields = 'name' | 'value' | 'checked' | 'onChange' | 'required'
/* We don't need to pass these props to form control anymore because we described it usages here */
export type FormModuleControlledFields = 'render' | 'defaultValue'

/* Groups from component props with form control props excluding props form controls */
export type FormModuleComponent<T, F> = React.FC<
  Omit<T, FormModuleComponentControlledFields> & Omit<FormControlProps<F>, FormModuleControlledFields>
>
/* Form module. Controlled components */
export type FormModule<F> = {
  Item: typeof FormItem
  Control: typeof FormControl
  Submit: typeof FormSubmit
  Input: FormModuleComponent<InputProps, F>
  Select: FormModuleComponent<SelectProps, F>
  Autocomplete: FormModuleComponent<AutocompleteProps, F>
  Checkbox: FormModuleComponent<CheckboxProps, F>
}
/* Form we get from loadable */
export type LoadableForm<F> = LoadableComponent<FormProps<F>> & FormModule<F>

/*
  Form module
  TODO Research. Dot notation doesn't work with hocs

  Callback needs for generic form logic
  const SignInForm = Form<SignInFormType>()
*/

const Form = <F,>(withLoadableFallbackOptions?: WithLoadableFallbackOptions): LoadableForm<F> => {
  const FormModule = withLoadableFallback<FormProps<F>>(
    import('./Form'),
    withLoadableFallbackOptions || {}
  ) as LoadableForm<F>

  /* Adding components to the module */
  FormModule.Control = FormControl
  FormModule.Item = FormItem
  FormModule.Submit = FormSubmit
  FormModule.Input = props => {
    const { name, validation, defaultValue, ...inputProps } = props
    const required = !!validation?.required

    return (
      <FormModule.Control
        name={name}
        validation={validation}
        defaultValue={defaultValue || ''}
        render={(props: ControllerRenderProps) => {
          const { value, ref, onChange } = props

          return <Input name={name} value={value} ref={ref} onChange={onChange} required={required} {...inputProps} />
        }}
      />
    )
  }

  FormModule.Input.displayName = 'FormInput'

  FormModule.Select = props => {
    const { name, validation, defaultValue, ...selectProps } = props
    const required = !!validation?.required

    return (
      <FormModule.Control
        name={name}
        validation={validation}
        defaultValue={defaultValue || null}
        render={(props: ControllerRenderProps) => {
          const { value, ref, onChange } = props

          return (
            <Select name={name} value={value} innerRef={ref} onChange={onChange} required={required} {...selectProps} />
          )
        }}
      />
    )
  }

  FormModule.Select.displayName = 'FormSelect'

  FormModule.Autocomplete = props => {
    const { name, validation, defaultValue, ...autocompleteProps } = props
    const required = !!validation?.required

    return (
      <FormModule.Control
        name={name}
        validation={validation}
        defaultValue={defaultValue || null}
        render={(props: ControllerRenderProps) => {
          const { value, ref, onChange } = props

          return (
            <Autocomplete
              name={name}
              value={value}
              innerRef={ref}
              onChange={onChange}
              required={required}
              {...autocompleteProps}
            />
          )
        }}
      />
    )
  }

  FormModule.Autocomplete.displayName = 'FormAutocomplete'

  FormModule.Checkbox = props => {
    const { name, validation, defaultChecked, ...checkboxProps } = props
    const required = !!validation?.required

    return (
      <FormModule.Control
        name={name}
        validation={validation}
        defaultValue={defaultChecked || false}
        render={(props: ControllerRenderProps) => {
          const { value, ref, onChange } = props

          return (
            <Checkbox
              name={name}
              checked={value}
              ref={ref}
              onChange={onChange}
              required={required}
              {...checkboxProps}
            />
          )
        }}
      />
    )
  }

  FormModule.Checkbox.displayName = 'FormCheckbox'

  return FormModule
}

export default Form
