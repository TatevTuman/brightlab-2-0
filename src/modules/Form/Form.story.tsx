import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import { action } from '@storybook/addon-actions'
import { storybook } from '@utils'
import FormModule from './Form.module'
import { FormProps } from './Form'

type TestForm = {
  input: string
  select: string
  autocomplete: string
  checkbox: boolean
}

const Form = FormModule<TestForm>()
const selectOptions = new Array(100).fill(0).map((_, index) => {
  return { label: 'Select Option' + index, value: index.toString() }
})
const autocompleteOptions = new Array(100).fill(0).map((_, index) => {
  return { label: 'Autocomplete Option' + index, value: index.toString() }
})

interface FormStoryTemplateProps extends FormProps<TestForm> {}

export default {
  title: 'Components/Form',
  argTypes: {
    onSubmit: { action: 'onSubmit' },
    onError: { action: 'onError' }
  }
} as Meta

const FormStoryTemplate: Story<FormStoryTemplateProps> = args => {
  return (
    <Form {...args}>
      {useFormMethods => {
        const { getValues } = useFormMethods

        return (
          <>
            <Form.Item>
              <Form.Input label={'Input'} name={'input'} />
            </Form.Item>
            {/*<Form.Item>*/}
            {/*  <Form.Select name={'select'} options={selectOptions} />*/}
            {/*</Form.Item>*/}
            {/*<Form.Item>*/}
            {/*  <Form.Autocomplete name={'autocomplete'} options={autocompleteOptions} />*/}
            {/*</Form.Item>*/}
            <Form.Item>
              <Form.Checkbox label={'Checkbox'} name={'checkbox'} />
            </Form.Item>
            <Form.Submit type={'primary'}>Submit</Form.Submit>
            <Form.Item>{JSON.stringify(getValues(), null, 2)}</Form.Item>
          </>
        )
      }}
    </Form>
  )
}

FormStoryTemplate.args = {
  onSubmit: action('onSubmit'),
  onError: action('onError')
}

export const FormStory = FormStoryTemplate.bind({})
FormStory.args = {
  ...FormStoryTemplate.args
}
