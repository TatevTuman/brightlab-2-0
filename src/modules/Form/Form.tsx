import React from 'react'
import {
  useForm,
  SubmitHandler,
  SubmitErrorHandler,
  UnpackNestedValue,
  DeepPartial,
  UseFormMethods,
  FieldName,
  FormProvider
} from 'react-hook-form'

import { Children } from '@types'
import { genericMemo } from '@hocs'
import './Form.scss'

export interface FormProps<F> {
  id?: string
  className?: string
  defaultValues?: UnpackNestedValue<DeepPartial<F>>
  children: (props: UseFormMethods<F>) => Children
  onSubmit: (form: UnpackNestedValue<F>, event?: React.BaseSyntheticEvent) => void
  onError: SubmitErrorHandler<F>
}

/*
  Form component that uses react-hook-form lib to manage form state without React state.

  Loading state is implemented via FormSubmit component that controls loading field like input in the form
  You can add here any logic you want to improve UX
*/

const Form = <F,>(props: FormProps<F>) => {
  const { id, className, defaultValues, onSubmit, onError, children } = props
  const useFormMethods = useForm<F>({ defaultValues })

  /* FormSubmit component controls loading field, so we can manage button state with react-hook-form */
  const setFormLoading = () =>
    setTimeout(
      () =>
        useFormMethods.setValue('loading' as FieldName<F>, false, {
          shouldValidate: false,
          shouldDirty: true
        }),
      500
    )

  /* Custom form submit handler */
  const handleFormSubmit: SubmitHandler<F> = async (form, event) => {
    await onSubmit(form, event)
    /* Set FormSubmit loading to false */
    setFormLoading()
  }

  /* Custom form error handler */
  const handleFormError: SubmitErrorHandler<F> = async (errors, event) => {
    await onError(errors, event)
    /* Set FormSubmit loading to false */
    setFormLoading()
  }

  return (
    <FormProvider {...useFormMethods}>
      <form
        id={id}
        className={'form ' + className}
        onSubmit={useFormMethods.handleSubmit<F>(handleFormSubmit, handleFormError)}
        noValidate
        data-testid={'form'}
      >
        {children(useFormMethods)}
      </form>
    </FormProvider>
  )
}

export default genericMemo(Form)
