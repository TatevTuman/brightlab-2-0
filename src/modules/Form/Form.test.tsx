import React from 'react'
import { render, cleanup, waitFor, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Autocomplete, Select } from '@components'
import { Button, Input } from '@elements'
import { RoleOptionValue } from '@types'
import { emailPattern } from '@utils'
import Form from './Form'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Form', () => {
  const props = {
    onSubmit: jest.fn(),
    onError: jest.fn(),
    defaultValues: {
      email: ''
    }
  }

  const form = () =>
    render(
      <Form<{ email: string; role: { name: string }; job: { name: string } }> {...props}>
        {useFormMethods => {
          return (
            <>
              <Input
                type={'email'}
                name={'email'}
                label={'Email'}
                validation={{
                  required: { value: true, message: 'Email is required' },
                  pattern: { value: emailPattern, message: 'It doesn`t seems to be an email' }
                }}
              />
              <Form.Submit>Submit</Form.Submit>
            </>
          )
        }}
      </Form>
    )

  it('renders correctly', async () => {
    const { container } = form()
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
  })

  it('submit on valid fields', async () => {
    const { getByText, getByLabelText } = form()

    await act(async () => {
      const email = await waitFor(() => getByLabelText('Email'))
      userEvent.type(email, 'test@test.com')
      const submit = await waitFor(() => getByText('Submit'))
      userEvent.click(submit)
    })

    expect(props.onError).toHaveBeenCalledTimes(0)
    expect(props.onSubmit).toHaveBeenCalledTimes(1)
  })
})
