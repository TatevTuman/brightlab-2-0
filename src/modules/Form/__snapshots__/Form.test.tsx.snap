// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<div>
  <form
    class="form undefined"
    novalidate=""
  >
    <div
      class="input"
      data-direction="vertical"
    >
      <label
        data-required="true"
        for="email"
      >
        Email
      </label>
      <div
        class="inputInner"
      >
        <input
          autocomplete="on"
          data-cursor="true"
          data-error="false"
          data-prefix="false"
          data-suffix="false"
          data-times="true"
          id="email"
          name="email"
          placeholder="Type a text"
          tabindex="0"
          type="email"
          value=""
        />
        <div
          class="inputInnerPrefix"
          data-prefix="false"
        />
        <div
          class="inputInnerSuffix"
          data-suffix="false"
        >
          <test-file-stub
            classname="inputInnerSuffixTimes"
          />
        </div>
      </div>
    </div>
    <button
      class="button button__primary-md formSubmit"
      data-centered="true"
      data-disabled="false"
      data-size="md"
      tabindex="0"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
